### PR TITLE
Implement QPE for operators

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**

Currently the [QPE](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.QuantumPhaseEstimation.html) template only supports operator pass-through as a matrix. It would be useful to be able to pass a quantum function and not have to think in arrays.

**Description of the Change:**

QPE is updated to take an `Operator` as the input unitary to perform phase estimation with.

**Benefits:**

Now we can perform QPE on a quantum operator

**Possible Drawbacks:**

This is a new feature added, which does not impact existing functionalities. However, it is worth noting that there are two potential approaches to performing phase estimation with a quantum function, specifically when constructing powers of the unitary. The first approach is to find the matrix representation of the circuit and calculate the powers numerically. 

This PR takes the second approach, which is to simply apply the quantum function multiple times. Admittedly, using this approach, the number of gate operations would scale exponentially and not linearly with the number of estimation bits, but I would argue that the second approach is closer to the intended implementation of QPE. 

When we analyze the theoretical complexity of QPE, we always treat the unitary as an oracle and discuss the complexity in terms of calls to the controlled unitary. I believe that the implementation here should reflect that, as opposed to taking the shortcut of calculating the matrix powers numerically.

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/2935
https://github.com/PennyLaneAI/pennylane/issues/3407

### CODE REVIEW

**Comment by albi3ro** [comment]
Hi @astralcai  and thanks for the pull request 🎉 

@KetpuntoG and I have clarified the problem description and the desired behaviour in the Issue.  You can see my comment on the problem description at: https://github.com/PennyLaneAI/pennylane/issues/2935#issuecomment-1318736193

**Comment by astralcai** [comment]
Hi @albi3ro,

I'm glad that we agree on using operations and not matrices when applying the controlled unitary.

This new call signature would definitely be a lot easier to implement, but there is one thing in my current implementation that I would like to keep. I imagine a list of operators can be defined on any set of wire labels, and once it is defined, it can be reused in many different places. Let's say someone decides to define a list of operations on wires 0 and 1, maybe we don't want to constrain the user to have to choose 0 and 1 as target wires in their QPE circuit (maybe their QPE is part of a larger circuit, and wires 0 and 1 are not even part of the QPE circuit). That is why I have this wire map in my code, let me know what you think of this.

Best,

Astral

**Comment by albi3ro** [comment]
Our new `qml.map_wires` function (which I see you are using 😄 ) acts on individual Operators, Measurement Processes, quantum scripts, quantum functions, or QNodes.  So we can still map wires on individual operators.

We can still define one `Operator`, and use it in multiple places with mapped wires through `qml.map_wires`.

I'm also wary of implicitly mapping wires under the hood.   I could see this leading to hard-to-diagnose unexpected behaviour. If the target wires are somehow described by the object provided by `unitary`, they should *only* be provided through `unitary`.

One thing we also need to start popularizing and advertising more, we can turn a list of operations into a `Prod` Operator.  For example,  the target operator could be:

```
qml.RX(x, 0) @ qml.RY(y, 1)
```
which would be equivalent to:
```
def qfunc():
    qml.RY(y, 1)
    qml.RY(x, 0)
```

As a final note, since the information on `target_wires` would be stored in the `unitary` operation, we would also need to override the `Operator.map_wires` method to update the wires on the base operation as well.

**Comment by astralcai** [comment]
Hi, how about if target wires are not provided, we use the wires defined in the unitary, but if target wires are explicitly provided, we do the wire map to the target wires? We can clearly explain this behaviour in the docstring of the QPE method.

**Comment by astralcai** [comment]
Hi @albi3ro,

I just made the update. Now the QPE can take a single operator or a list of operators as the unitary with no target_wires specified. However, I kept the logic where if target wires are specified in addition to the operators, I would still map the operators to the target wires. 

Best,

Astral

**Comment by albi3ro** [comment]
### Switching the order of arguments in the call signature:
Switching `target_wires` and `estimation wires` in the call signature would be a breaking change that could lead to a lot of hard-to-diagnose problems.  Instead, we can go with the simpler solution: giving both `target_wires` and `estimation_wires` default values of `None`.  While `estimation_wires` would then need to be provided as a keyword argument, that is perfectly reasonable restriction.
```
QPE(op, estimation_wires=[1,2,3])
```
This leads to the call signature:
```
QPE(unitary: Union[ndarray, Operator], target_wires=None, estimation_wires=None, do_queue=True, id=None)
```

### Mapping the wires internally

Deciding whether or not to map the wires internally hinges on the answers to two questions:

1) How hard is it to let the user control mapping the wires themselves?
2) What is the probability of bugs and unexpected behavior if we map the wires internally? How hard would these problems be to debug?

With the release of `qml.map_wires`, users can now easily map the wires themselves. 

Even if we document the behavior of the internal wire mapping, not everyone will not read the docstring closely enough. If wires are mistakenly mapped, this could be tricky to diagnose. I'd rather preemptively avoid those situations.

We should not map the wires internally. Instead, if `unitary` is an `Operator` and `target_wires` is specified, we should raise an error.

### Accepting a list of Operators

We will already be overloading the `unitary` keyword with both `array` and `Operator`. Including `list[Operator]` will increase the complexity of the interface and behavior with only minimal improvements to functionality. As we start expanding the use of `qml.prod` and working with the product of operators, the use cases for `list[Operator]` will decrease even more.

We should not accept a list of operators.  That adds too much complexity to the interface.

### Map wires method

You can check out the source code for [`Controlled.map_wires`](https://github.com/PennyLaneAI/pennylane/blob/ffa6e46aae3ab1c94c8c9c8d2761a900ca91abd0/pennylane/ops/op_math/controlled_class.py#L276) for an example implementation.  This instance method should create a  new operation that is a copy except for mapped wires. `Controlled` is another class that has a target operation and additional "control"/ "estimation" wires.

Though we need to update the method to touching the private `_wires` attribute.  Just recursively call `op.map_wires(map)` isntead.

**Comment by astralcai** [comment]
Hi @albi3ro,

I am trying to implement QPE using the ```qml.pow``` operator as you suggested, but have encountered an issue. If the unitary is specified as a ```Prod```, such as ```qml.RX(param, wires=[0]) @ qml.CNOT(wires=[0, 1])```, then during decomposition of ```(qml.RX(param, wires=[0]) @ qml.CNOT(wires=[0, 1])) ** 1```, we call ```Pow.decomposition()```, which then calls ```Prod.pow``` (line 283 of ```ops/op_math/pow.py```, where ```self.base``` is a ```Prod``` as defined earlier). However, since ```Prod.pow``` is not defined, it defaults to ```Operator.pow```, which returns ```[copy.copy(self)]``` (line 1041 of ```operations.py```), but this is not queued. 

As a result, in ```Operator.expand()```, when we first create a tape (line 1094 of ```operations.py```), and call ```self.decomposition()``` (line 1100), the result of that decomposition is not added to the tape, which leads to the final tape simply missing the single power of the unitary.

I imagine we can override ```Prod.pow``` with the same logic as ```Operator.pow```, except that we return ```[qml.apply(self)]``` instead of ```[copy.copy(self)]``` if ```QueuingManager.recording()```, similar to the logic in ```Pow.decomposition``` (line 286 of ```ops/op_math/pow.py```). What do you suggest?

Best,

Astral

**Comment by albi3ro** [comment]
@astralcai  This appears to be a bug.  Thank you for finding it. Would you mind opening a bug report issue on this? That way we can track what bugs exist and were fixed.

This problem extends far beyond `Prod`, and can be cleanly described by this:

```pycon
>>> qml.pow(qml.PauliX(0), 1).expand().circuit
[]
>>> qml.pow(qml.PauliX(0), 1).decomposition()
[PauliX(wires=[0])]
```

Basically, `Operator.pow` needs to queue all its constituents. As you stated, this can be solved by using `apply` instead of `copy` in the default behaviour.

**Comment by astralcai** [comment]
Hi @albi3ro,

If I simply update ```Operator.pow``` to use ```apply``` instead of ```copy```, the issue would be fixed. Do I still need to open a bug report in this case?

Best,

Astral

**Comment by albi3ro** [comment]
Yes please.  It helps us keep track of what things have required bug fixes. A bug fix would also require dedicated tests and a changelog entry.  

I've also realized we will need to override the `queue` method as well, see [`SymbolicOp.queue`](https://github.com/PennyLaneAI/pennylane/blob/6881ca8076983f4f324492cfeb8edb10ec06a5be/pennylane/ops/op_math/symbolicop.py#L112) for an example.  `SymbolicOp` is the base class for "operator that contains another operator", which is what `QPE` will now be.  Basically, if the target operation is constructed inside the quantum function, we don't want it to end up in the final circuit, so we need to make the target operation as "owned".

**Comment by astralcai** [comment]
> Yes please. It helps us keep track of what things have required bug fixes. A bug fix would also require dedicated tests and a changelog entry.
> 
> I've also realized we will need to override the `queue` method as well, see [`SymbolicOp.queue`](https://github.com/PennyLaneAI/pennylane/blob/6881ca8076983f4f324492cfeb8edb10ec06a5be/pennylane/ops/op_math/symbolicop.py#L112) for an example. `SymbolicOp` is the base class for "operator that contains another operator", which is what `QPE` will now be. Basically, if the target operation is constructed inside the quantum function, we don't want it to end up in the final circuit, so we need to make the target operation as "owned".

Makes sense, I will add the `queue` method override.

File: doc/releases/changelog-dev.md
Comment (albi3ro) [inline doc/releases/changelog-dev.md:75]: We don't need an example in the changelog for this case.  We want to restrict examples to big ticket items where we want to draw a lot of attention. Otherwise, people can just read docstrings.

File: doc/releases/changelog-dev.md
Comment (albi3ro) [inline doc/releases/changelog-dev.md:50]:
```suggestion
* QPE can now accept a target operator instead of a matrix and target wires pair.
  [(#3373)](https://github.com/PennyLaneAI/pennylane/pull/3373)
```
This entry should also go under "Improvements" instead of "New Features".

File: pennylane/templates/subroutines/qpe.py
Comment (albi3ro) [inline pennylane/templates/subroutines/qpe.py:113]:
```suggestion
        with target wires, the target wires should not be provided for the QPE.
```

File: pennylane/templates/subroutines/qpe.py
Comment (albi3ro) [inline pennylane/templates/subroutines/qpe.py:156]: To avoid having to check between the two cases (matrix versus operator) downstream, we could create a `QubitUnitary` operator here to unify the two cases.

Then `unitary` would always be a `Operator` after initialization.  We could then store the operator in the hyperparameters.

This would also mean we would need a data getter and setter like in [`SymbolicOp`](https://github.com/PennyLaneAI/pennylane/blob/6881ca8076983f4f324492cfeb8edb10ec06a5be/pennylane/ops/op_math/symbolicop.py#L79). At least if we still want to treat the `QPE` data like the target operator data.

File: pennylane/templates/subroutines/qpe.py
Comment (astralcai) [reply pennylane/templates/subroutines/qpe.py:113]: Oops

File: pennylane/templates/subroutines/qpe.py
Comment (astralcai) [reply pennylane/templates/subroutines/qpe.py:156]: Good idea! I will implement it now.

Comment (astralcai) [reply pennylane/templates/subroutines/qpe.py:156]: If we wrap the unitary matrix in a `QubitUnitary`, then the controlled unitary would be implemented with `qml.ctrl` which returns a general `Controlled` operator instead of `ControlledQubitUnitary`. This is completely fine I guess, but the `ControlledQubitUnitary` class would seem a bit redundant after this change.

Comment (albi3ro) [reply pennylane/templates/subroutines/qpe.py:155]: Given `ControlledQubitUnitary` probably has less features than `ControlledOp`, I'm not too concerned.  We'll probably be replacing most of it soon anyway.

File: pennylane/templates/subroutines/qpe.py
Comment (Qottmann) [inline pennylane/templates/subroutines/qpe.py:116]:
```suggestion
```

File: tests/templates/test_subroutines/test_qpe.py
Comment (albi3ro) [inline tests/templates/test_subroutines/test_qpe.py:175]:
```suggestion
            tape = qml.tape.QuantumScript([qml.QuantumPhaseEstimation(unitary, estimation_wires=estimation_wires)],
                [qml.probs(estimation_wires)], prep = [ qml.QubitStateVector(eig_vec, wires=target_wires)])
```
We're trying to move away from using `QuantumTape` internally.  Not deprecated yet, but this is now the recommended way of doing things.

**Review by albi3ro — COMMENTED** [review]
Almost there!

The last remaining change I can see is moving `unitary` to be a hyperparameter instead of a parameter/ data.  Since `unitary` is an Operator instead of numbers, it should be in `hyperparameters`.

We have a feature freeze for v0.28 starting on Friday, but I think we can get this PR into the release.

We're moving away from recording with a `QuantumTape` to directly constructing a `QuantumScript`, so I made some suggestions to the tests, but those do not block this PR.

Just wanted to share how nice drawing out circuits with this can be now:

`qml.draw_mpl(circuit, expansion_strategy="device')()`
image

Much easier to see the structure of the algorithm ;)

File: pennylane/templates/subroutines/qpe.py
Comment (albi3ro) [inline pennylane/templates/subroutines/qpe.py:186]:
```suggestion
```

File: pennylane/templates/subroutines/qpe.py
Comment (albi3ro) [inline pennylane/templates/subroutines/qpe.py:221]:
```suggestion
        wires, unitary, target_wires, estimation_wires
```

File: pennylane/templates/subroutines/qpe.py
Comment (albi3ro) [inline pennylane/templates/subroutines/qpe.py:231]:
```suggestion
```

**Review by albi3ro — COMMENTED** [review]
By getting rid of `num_params = 1`, we can just not have any data.

Not a strictly blocking request, so I'll approve it tomorrow anyway if this change isn't made by then.

**Comment by astralcai** [comment]
@albi3ro 

I made the update. One test failed, but the failing test case has nothing to do with QPE, and I can't even replicate the fail on my own machine.

**Review by albi3ro — APPROVED** [review]
@astralcai I updated with master, and we'll see if that fixes the CI problem. If not, I'll debug it myself this afternoon.

Thanks for your work on this PR!

**Comment by astralcai** [comment]
> @astralcai I updated with master, and we'll see if that fixes the CI problem. If not, I'll debug it myself this afternoon.
> 
> Thanks for your work on this PR!

@albi3ro seems that the issue is gone now. I just merged from master again. I'm actually going out this afternoon and won't be able to work on this for the rest of the day. Hopefully you can have it merged. Thanks for all the support :)

**Pull request merged**
